lib/elasticity/instance_fleet.rb

```diff
@@ -0,0 +1,121 @@
+module Elasticity
+
+  class InstanceFleet
+
+    MAX_NUM_INSTANCE_TYPES = 5
+    ROLES = %w(MASTER CORE TASK)
+    TIMEOUT_ACTIONS = %w(SWITCH_TO_ON_DEMAND TERMINATE_CLUSTER)
+    MIN_PROVISIONING_TIMEOUT_MINUTES = 5
+    MAX_PROVISIONING_TIMEOUT_MINUTES = 1440
+    VALID_BLOCK_DURATION_MINUTES = [60, 120, 180, 240, 300, 360]
+
+    attr_accessor :role
+    attr_accessor :name
+    attr_accessor :target_on_demand_capacity
+    attr_accessor :target_spot_capacity
+    attr_accessor :spot_block_duration_minutes
+    attr_accessor :spot_provisioning_timeout_minutes
+    attr_accessor :spot_provisioning_timeout_action
+
+    def initialize
+      @instance_type_configs = []
+      @role = 'CORE'
+      @target_on_demand_capacity = 0
+      @target_spot_capacity = 0
+    end
+
+    def check_capacity(capacity)
+      if capacity <= 0
+        raise ArgumentError, "Instance fleets require capacity at least 1! (#{capacity} requested)"
+      end
+      if @role == 'MASTER' && capacity != 1
+        raise ArgumentError, "MASTER instance fleets can only have capacity 1! (#{capacity} requested)"
+      end
+    end
+
+    def target_on_demand_capacity=(capacity)
+      check_capacity(capacity)
+      @target_on_demand_capacity = capacity
+    end
+
+    def target_spot_capacity=(capacity)
+      check_capacity(capacity)
+      @target_spot_capacity = capacity
+    end
+
+    def role=(group_role)
+      if !ROLES.include?(group_role)
+        raise ArgumentError, "Role must be one of #{ROLES.join(', ')} (#{group_role} was requested)."
+      end
+      @role = group_role
+    end
+
+    def name=(group_name)
+      @name = group_name
+    end
+
+    def spot_provisioning_timeout_action=(action)
+      if !TIMEOUT_ACTIONS.include?(action)
+        raise ArgumentError, "Timeout action must be one of #{TIMEOUT_ACTIONS.join(', ')} (#{action} was requested)."
+      end
+      @spot_provisioning_timeout_action = action
+    end
+
+    def spot_provisioning_timeout_minutes=(timeout_minutes)
+      if !(MIN_PROVISIONING_TIMEOUT_MINUTES..MAX_PROVISIONING_TIMEOUT_MINUTES).include?(timeout_minutes)
```

Comment:
I am wondering if this is faster/simpler than `timeout_minutes >= MIN_PROVISIONING_TIMEOUT_MINUTES and  timeout_minutes <= MAX_PROVISIONING_TIMEOUT_MINUTES`

Reply:
I considered this. I would have done it if Ruby supported chained inequalities, like Python. This struck me as the cleanest way. I highly doubt there is any detectable difference in performance given that we do this ~once per cluster.

Reply:
.between? https://ruby-doc.org/core-2.2.0/Comparable.html#method-i-between-3F

Reply:
Yes!! ^